Added large fluid extractor multi

RecursivePineapple opened the pull request:
This PR adds a fluid extractor multi, to help with replacing the LPF.

The multi uses the currently unused solenoid coils. Increasing the tier of these coils gives 2 extra parallels.
The multi also uses heating coils. Each coil tier adds 10% extra speed (additive).

By default, the multi has a 20% speed bonus and a 20% EU reduction (120% speed total and 80% EU/t total).

![2024-08-04_03 04 00](https://github.com/user-attachments/assets/64859434-31f7-4f2c-a062-d23a4948fca8)

RecursivePineapple commented:
There's a bug with the structure check, but I'm going to be away for a few days on a trip. If someone is impatient and wants it done sooner, feel free to work on it.

MCTBL commented:
You sure this mutli is for replace LPF? this mutli even worse than LPF, and u sure only +2 paralles per tier is enough for late game?

tommy307507 commented:
I am not familiar with the solenoid coil tiering, maybe you should provide a table of the tier of heating coils and solenoid coils with respect to voltage tier V.S. the LPF if you intend to replace the multi. From your current description the coils have 11 tiers and max out on 22parallels in UIV, I don't think this gets close to LPF. As a zpm player, I will prefer the PA than this machine. This needs more parallels and more reviews from players with late game experience.

RecursivePineapple commented:
I've changed the parallel, speed, and EU/t mechanics.
Now, each coil tier gives +8 parallels per tier (for a total of 96 with UMV coils) along with a 25% multiplicative EU/t penalty.
Each heating coil tiers gives +10% speed and a 5% EU/t reduction to balance the cost from the solenoid coils.
This should let early players optimize their setups - one for high parallels but low EU/t recipes, and one for low parallel high EU/t recipes. In the late game this shouldn't matter much.

Here's a recipe that takes 2 EU/t and 15 seconds.
![image](https://github.com/user-attachments/assets/3cf1ecd7-487a-46c8-a906-138fb3039180)
![image](https://github.com/user-attachments/assets/ed221320-c436-462e-ade6-40bb4a4fc03b)

Hikari1nVoid commented:
So This new multi gives 2.3x Speed Boost at the cost of 5.6x power comsumption with TOP-TIER coil.I don't see a single point for coil upgrade now.

MCTBL commented:
as what u said the final eu cost should equals this equation
`(1.25^coil tier) * (1-0.05 * heat coil tier)`
and this is your code
``` java
    public double getEUMultiplier() {
        double heatingBonus = 1 / (1 + mCoilLevel.getTier() * HEATING_COIL_EU_BONUS);
        double solenoidPenalty = Math.pow(1 + SOLENOID_EU_PENALTY, (double) mSolenoidLevel);

        return solenoidPenalty * heatingBonus;
    }
```
and this is what your code mean
`[1/(1+heat coil tier * 0.05)] * (1.25^coil tier)`
which one is true?
Please correct me if I'm wrong.

RecursivePineapple commented:
I'm going to undraft this because there's no complaints with the current mechanics. It might still be too overpowered, but it's also just a fluid extractor. If it's only a bit OP, I'll just make the higher tier solenoids expensive since that's where most of the speedup comes from.

RecursivePineapple commented:
@Hikari1nVoid Solenoid upgrades increase power usage by a bit but increase parallels significantly. Higher tiers increase the power usage more. Heating coils give a slight speed boost and a power discount.

RecursivePineapple commented:
@MCTBL They should be equivalent. I'll double check, but if the heating coil tier is 2 (MV) then the EU multiplier should be `1 / 1.1`. If the solenoid tier is 2 (MV) then the EU penalty should be `1.25 ^ 2` which is 1.5625. If you multiply the two together you get a ~1.42x multiplier total.
This combination should get you 16 parallels and a 20% speed boost in exchange for the power increase.

MCTBL commented:
> @MCTBL They should be equivalent. I'll double check, but if the heating coil tier is 2 (MV) then the EU multiplier should be `1 / 1.1`. If the solenoid tier is 2 (MV) then the EU penalty should be `1.25 ^ 2` which is 1.5625. If you multiply the two together you get a ~1.42x multiplier total. This combination should get you 16 parallels and a 20% speed boost in exchange for the power increase.


  | (1-0.05\*coil) | 1/(1+coil\*0.05) | (1-0.05\*coil)-1/(1+coil\*0.05)
-- | -- | -- | --
0 | 1 | 1 | 0
1 | 0.95 | 0.952380952 | -0.002380952
2 | 0.9 | 0.909090909 | -0.009090909
3 | 0.85 | 0.869565217 | -0.019565217
4 | 0.8 | 0.833333333 | -0.033333333
5 | 0.75 | 0.8 | -0.05
6 | 0.7 | 0.769230769 | -0.069230769
7 | 0.65 | 0.740740741 | -0.090740741
8 | 0.6 | 0.714285714 | -0.114285714
9 | 0.55 | 0.689655172 | -0.139655172
10 | 0.5 | 0.666666667 | -0.166666667
11 | 0.45 | 0.64516129 | -0.19516129
12 | 0.4 | 0.625 | -0.225
13 | 0.35 | 0.606060606 | -0.256060606
14 | 0.3 | 0.588235294 | -0.288235294

will, i don't think they *should* be equivalent

Hikari1nVoid commented:
<img width="1079" alt="image" src="https://github.com/user-attachments/assets/d76b58f3-9f5d-4356-aeac-262a370b96c3">
According to the fomula provided above,here is the graph for coil tier-EU-multiplier.
Also is 96x parallel significant for late tiers...?

MCTBL commented:
which voltage tier do u reach?

chochem commented on src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeFluidExtractor.java:
```diff
@@ -0,0 +1,433 @@
+        return solenoidPenalty * heatingBonus;
+    }
+
+    private class LFEProcessingLogic extends ProcessingLogic {
```
why are you not just overriding createProcessingLogic() like every other multi does?

chochem replied:
all you need here is like
`@Override
    protected ProcessingLogic createProcessingLogic() {
        return new ProcessingLogic().setSpeedBonus(...)
            .setEuModifier(...)
            .setMaxParallelSupplier(this::getParallels);
    }`

RecursivePineapple replied:
Sorry, that was old code from when I did `recipe tier - solenoid/heating coil tier` (I remove that mechanic because it was stupid). I'll clean it up.

chochem commented:
> Also is 96x parallel significant for late tiers...?

@Hikari1nVoid this is a IV multi. 8xtier is pretty normal and sounds good. 

> They should be equivalent.

@RecursivePineapple  (1-nx) and 1/(1+nx) are most definetely not the same function. I think they just wanted you to say that you mean 1/(1+nx).
I guess technically one might call that a +5% energy efficiency per tier instead of a 5% energy discount per tier which kinda is the other formula and thus why they were confused about it.

Hikari1nVoid commented:
> > Also is 96x parallel significant for late tiers...?
> 
> @Hikari1nVoid this is a IV multi. 8xtier is pretty normal and sounds good.
> 
Tbf despite this is a IV multi , there are no more multis at the endgame doing fluid extractor workload.Really bored to see a wall of LPF but at this parellel count this will end up being another boring Multiblock wall.(and it took more space than current LPF:(

RecursivePineapple commented:
> (1-nx) and 1/(1+nx) are most definetely not the same function
@chochem my algebra must be getting rusty, I'll fix it later today

chochem commented:
> > (1-nx) and 1/(1+nx) are most definetely not the same function
> 
> @chochem my algebra must be getting rusty, I'll fix it later today

it does not really need a fix. if you want your formula I dont see an issue with it.

RecursivePineapple commented:
> Tbf despite this is a IV multi , there are no more multis at the endgame doing fluid extractor workload

@Hikari1nVoid I know the devs have plans to add better multis that scale in the latter tiers (stargate doubly so). They don't want IV multis to scale that far, so I'm fully expecting this to get nerfed at/before those tiers.
I can work on the late game / pseudo-mega multi as well, but it'll have to wait because I have other projects I want to finish first.

Connor-Colenso commented:
> > > Also is 96x parallel significant for late tiers...?
> 
> > 
> 
> > @Hikari1nVoid this is a IV multi. 8xtier is pretty normal and sounds good.
> 
> > 
> 
> Tbf despite this is a IV multi , there are no more multis at the endgame doing fluid extractor workload.Really bored to see a wall of LPF but at this parellel count this will end up being another boring Multiblock wall.(and it took more space than current LPF:( )
> 
> 

The answer to this imo is later game better multis. Not to make sure this stretches all the way to stargate.

NotAPenguin0 commented on src/main/java/com/github/bartimaeusnek/bartworks/API/BorosilicateGlass.java:
Not sure if these changes belong in this PR, but if it all works and doesn't conflict with any possible refactors done on master since then it's probably fine.

NotAPenguin0 commented on src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeFluidExtractor.java:
```diff
@@ -0,0 +1,410 @@
+    @Nullable
+    private Byte mSolenoidLevel = null;
+    private int mCasingAmount;
+    private boolean mBadStructure = false, mStructureBadGlassTier = false, mStructureBadCasingCount = false;
```
Not sure you need a `mBadStructure` boolean as a member variable. `MultiBlockBase::mMachine` is exactly this.

NotAPenguin0 commented on src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeFluidExtractor.java:
```diff
@@ -0,0 +1,410 @@
+        tt.addMachineType("Fluid Extractor")
+            .addInfo("Controller block for the Large Fluid Extractor")
+            .addInfo(String.format(
+                "%d%% the speed of single block machines of the same voltage",
```
Use consistent formatting with the rest of the multiblocks: 200% base speed is written as "100% faster than singleblock machines of the same voltage".

NotAPenguin0 commented on src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeFluidExtractor.java:
```diff
@@ -0,0 +1,410 @@
+
+        data.addAll(Arrays.asList(super.getInfoData()));
+
+        if (mBadStructure) {
```
`mMachine`.

NotAPenguin0 commented on src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeFluidExtractor.java:
```diff
@@ -0,0 +1,410 @@
+
+            data.add(
+                String.format(
+                    "%sEnergy hatch tier (%d) is too high for the glass tier (%d).%s",
```
In my opinion, these checks should not be in the scanner output but in the multiblock GUI, displayed as a reason for structure failing or shutting down, though there's not really a good API for doing that right now other than setting a custom `ShutDownReason`

RecursivePineapple replied:
I added a gui text widget to fix this and removed these scanner lines. It looks a lot better, thanks for the suggestion.

RecursivePineapple commented:
Here are the new stats. I changed the mechanics significantly because I couldn't balance the old ones properly.
![image](https://github.com/user-attachments/assets/f9bb0a13-0def-47d0-84c4-d629851999a1)

Here's the new tooltip. I changed pretty much every number here.
![2024-08-27_15 07 42](https://github.com/user-attachments/assets/a0ff1961-624a-4bdf-a98c-0aed3370b660)
